### PR TITLE
[SPARK-31526][SQL][TESTS] Add a new test suite for ExpressionInfo

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This PR intends to add a new test suite for `ExpressionInfo`. Major changes are as follows;

 - Added a new test suite named `ExpressionInfoSuite`
 - To improve test coverage, added a test for error handling in `ExpressionInfoSuite`
 - Moved the `ExpressionInfo`-related tests from `UDFSuite` to `ExpressionInfoSuite`
 - Moved the related tests from `SQLQuerySuite` to `ExpressionInfoSuite`
 - Added a comment in `ExpressionInfoSuite` (followup of https://github.com/apache/spark/pull/28224)

### Why are the changes needed?

To improve test suites/coverage.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added tests.

### CODE REVIEW

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java [hunk 1]
```diff
@@ -39,6 +39,9 @@
     private String since;
     private String deprecated;
 
+    // The groups below are almost the same with ones defined in `sql/functions.sql`.
+    // But, `collection_funcs` is split into fine-grained three groups: `array_funcs`, `map_funcs`,
+    // and `json_funcs`.
```
Comment (maropu) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java:44]: This is a minor follow-up for #28224, @HyukjinKwon

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    val errMsg = intercept[IllegalArgumentException] {
+      val invalidGroupName = "invalid_group_funcs"
+      new ExpressionInfo("testClass", null, "testName", null, "", "", "", invalidGroupName, "", "")
+    }.getMessage
+    assert(errMsg.contains("'group' is malformed in the expression [testName]."))
+  }
+
+  test("error handling in ExpressionInfo") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:60]: To improve test coverage, a new test added here.

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.catalyst.FunctionIdentifier
+import org.apache.spark.sql.catalyst.expressions.ExpressionInfo
+import org.apache.spark.sql.execution.HiveResult.hiveResultString
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSparkSession
+
+class ExpressionInfoSuite extends SparkFunSuite with SharedSparkSession {
+
+  test("Replace _FUNC_ in ExpressionInfo") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:31]: from `UDFSuite`

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    assert(info.getName === "upper")
+    assert(info.getClassName === "org.apache.spark.sql.catalyst.expressions.Upper")
+    assert(info.getUsage === "upper(str) - Returns `str` with all characters changed to uppercase.")
+    assert(info.getExamples.contains("> SELECT upper('SparkSql');"))
+    assert(info.getSince === "1.0.1")
+    assert(info.getNote === "")
+    assert(info.getExtended.contains("> SELECT upper('SparkSql');"))
+  }
+
+  test("group info in ExpressionInfo") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:42]: from `UDFSuite`

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    assert(errMsg2.contains("'since' is malformed in the expression [testName]."))
+
+    val errMsg3 = intercept[IllegalArgumentException] {
+      val invalidDeprecated = "  invalid deprecated"
+      new ExpressionInfo("testClass", null, "testName", null, "", "", "", "", "", invalidDeprecated)
+    }.getMessage
+    assert(errMsg3.contains("'deprecated' is malformed in the expression [testName]."))
+  }
+
+  test("using _FUNC_ instead of function names in examples") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:80]: from `SQLQuerySuite`

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        if (!exprExamples.isEmpty && !ignoreSet.contains(className)) {
+          assert(exampleRe.findAllIn(exprExamples).toIterable
+            .filter(setStmtRe.findFirstIn(_).isEmpty) // Ignore SET commands
+            .forall(_.contains("_FUNC_")))
+        }
+      }
+    }
+  }
+
+  test("check outputs of expression examples") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:107]: from `SQLQuerySuite`

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+      val clonedSpark = spark.cloneSession()
+      // Coalescing partitions can change result order, so disable it.
+      clonedSpark.sessionState.conf.setConf(SQLConf.COALESCE_PARTITIONS_ENABLED, false)
+      val info = clonedSpark.sessionState.catalog.lookupFunctionInfo(funcId)
+      val className = info.getClassName
+      if (!ignoreSet.contains(className)) {
+        withClue(s"Function '${info.getName}', Expression class '$className'") {
+          val example = info.getExamples
+          checkExampleSyntax(example)
+          example.split("  > ").toList.foreach {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:143]: minor: I modified this line a bit as my IDE suggested.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java [hunk 8]
```diff
@@ -39,6 +39,9 @@
     private String since;
     private String deprecated;
 
+    // The groups below are almost the same with ones defined in `sql/functions.sql`.
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java:42]: maybe you meant `functions.sql` -> `functions.scala`?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java (hunk 1 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java:44]: Nice but what about we document this in `ExpressionDescription`, and here we say things like please refer `ExpressionDescription`?

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java (hunk 8 above)
Comment (maropu) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java:42]: Ah, yes...

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java:44]: Yea, the idea looks nice. I'll update.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java [hunk 9]
```diff
@@ -103,6 +103,12 @@
     String arguments() default "";
     String examples() default "";
     String note() default "";
+    /**
+     * Valid group names are almost the same with one defined as `groupname` in
+     * `sql/functions.scala`. But, `collection_funcs` is split into fine-grained three groups:
+     * `array_funcs`, `map_funcs`, and `json_funcs`. See `ExpressionInfo` for the
```
Comment (yaooqinn) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java:109]: Why all the csv _funcs are left behind?

Comment (maropu) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java:109]: This is just a basic set for the 3.0 release. Yea, we can improve docs by adding more groups. Actually, we need to assign all the exprs to groups for removing the SQL Built-in Function doc:
https://github.com/apache/spark/pull/28224#issuecomment-616903015 @HyukjinKwon

Comment (yaooqinn) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java:109]: I see, thanks

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.0.

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala (hunk 6 above)
Comment (advancedxy) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:107]: Hi, @maropu this test case has a minor problem when running individually for machines in different timezones: expressions such as `FromUnixTime` are timezone aware. 
In `SQLQuerySuite`, the timezone is set explicitly in `org.apache.spark.sql.QueryTest`, however `ExpressionInfoSuite` doesn't set timezone, thus fails this test case.

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:107]: @advancedxy, that was fixed SPARK-31725.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:107]: Have you checked the thread in https://github.com/apache/spark/pull/28538?

Comment (advancedxy) [reply sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:107]: @HyukjinKwon and @maropu thanks for the information, I was checking an older version of Spark 3.0

**Pull request closed without merging**